File: src/java/com/codecool/gameoflife/view/GameLoop.java
package com.codecool.gameoflife.view;

import javafx.animation.AnimationTimer;

public class GameLoop extends AnimationTimer {

    private long previousTime = System.nanoTime();
    private long delayInMs;
    private FrameAction action;

    GameLoop(int delayInMs, FrameAction action) {
        this.delayInMs = delayInMs * 1_000_000; // to nano seconds
        this.action = action;
    }

    public void setDelay(int delayInMs) {
        this.delayInMs = delayInMs * 1_000_000;
    }

    @Override
    public void handle(long now) {
        if (now - previousTime < delayInMs) {
            return;
        }
        action.run();
        previousTime = now;
    }
}

File: src/kotlin/com/codecool/gameoflife/view/GameLoop.kt
package com.codecool.gameoflife.view

import javafx.animation.AnimationTimer

class GameLoop(delayInMs: Int, private val action: Runnable) : AnimationTimer() {

    private var previousTime = System.nanoTime()
    private var delay = delayToNano(delayInMs)

    fun setDelay(delayInMs: Int) {
        this.delay = delayToNano(delayInMs)
    }

    private fun delayToNano(delayInMs: Int) = (delayInMs * 1000000).toLong()

    override fun handle(now: Long) {
        if (now - previousTime > delay) {
            action.run()
            previousTime = now
        }
    }
}
